### ISSUE TITLE
Version validation prevents automation tools from executing extension update statement

### ISSUE THREAD

**Issue by Envek** [postgrespro/pg_pathman#95]
Hi everyone!

I'm trying to automate pg_pathman upgrading with Puppet.

So, I'm trying to do so:

 1. Clone/checkout required version from git to temp repo
 2. If revision in repo changed, execute `make install USE_PGXS=1`
 3. If build succeeded, restart PostgreSQL
 4. Check installed pg_pathman version in target database via `SELECT * FROM pg_extension WHERE extname = 'pg_pathman' AND extversion = '1.3';`
 5. If previous query returned no tuples (target version not installed), upgrade it via `ALTER EXTENSION pg_pathman UPDATE TO "1.3"; SET pg_pathman.enable = t`

The problem is that after upgrade from 1.2 to 1.3 if run any command in psql it will fail with following error:

    ОШИБКА:  pg_pathman's Pl/PgSQL frontend is incompatible with its shared library
    DETAIL:  consider performing an update procedure
    HINT:  pg_pathman will be disabled to allow you to resolve this issue

And as official Puppet's PostgreSQL module uses `psql` for executing commands, it can't even upgrade pg_pathman as any command passed to psql will fail immediately, and manual intervention is required to finish upgrade (because I'm only human after all and able to guess to execute command one more time):

```
psql (9.5.5)
Type "help" for help.

smp=# SELECT * FROM pg_extension WHERE extname = 'pg_pathman' AND extversion = '1.3';
ОШИБКА:  pg_pathman's Pl/PgSQL frontend is incompatible with its shared library
DETAIL:  consider performing an update procedure
HINT:  pg_pathman will be disabled to allow you to resolve this issue

smp=# SELECT * FROM pg_extension WHERE extname = 'pg_pathman' AND extversion = '1.3';
 extname | extowner | extnamespace | extrelocatable | extversion | extconfig | extcondition
---------+----------+--------------+----------------+------------+-----------+--------------
```

```
psql (9.5.5)
Type "help" for help.

smp=# ALTER EXTENSION pg_pathman UPDATE TO "1.3";
ОШИБКА:  pg_pathman's Pl/PgSQL frontend is incompatible with its shared library
DETAIL:  consider performing an update procedure
HINT:  pg_pathman will be disabled to allow you to resolve this issue
smp=# ALTER EXTENSION pg_pathman UPDATE TO "1.3";
ПРЕДУПРЕЖДЕНИЕ:  Don't forget to execute "SET pg_pathman.enable = t" to activate pg_pathman
ALTER EXTENSION
smp=# SET pg_pathman.enable = t;
ЗАМЕЧАНИЕ:  RuntimeAppend, RuntimeMergeAppend and PartitionFilter nodes and some other options have been enabled
SET
smp=# SELECT * FROM pg_extension WHERE extname = 'pg_pathman' AND extversion = '1.3';
  extname   | extowner | extnamespace | extrelocatable | extversion |       extconfig       | extcondition
------------+----------+--------------+----------------+------------+-----------------------+--------------
 pg_pathman |    16384 |    116179045 | f              | 1.3        | {116179175,116179185} | {"",""}
```

Can you please change **error** to some kind of **warning** that will be visible to humans, but will not prevent programs from executing statements?

Thank you in advance

**Comment by Envek** [comment]
**Expected behaviour**

This command executes successfully (exit code 0), prints no tuples when pg_pathman 1.2 active in target_db and pg_pathman 1.3 installed, PostgreSQL instance restarted:
```sh
psql target_db -t -c "SELECT * FROM pg_extension WHERE extname = 'pg_pathman' AND extversion = '1.3';"
```

This command executes successfully too (and upgrades pg_pathman):
```sh
psql target_db -t -c 'ALTER EXTENSION pg_pathman UPDATE TO "1.3"; SET pg_pathman.enable = t'
```

After that first command returns 1 tuple:

```
$ psql target_db -t -c "SELECT * FROM pg_extension WHERE extname = 'pg_pathman' AND extversion = '1.3';"
 pg_pathman |    16384 |    116179045 | f              | 1.3        | {116179175,116179185} | {"",""}

$ echo $?
0
```

**Actual behaviour**

These commands fails with error with exit code 1 and does nothing:

```
$ psql target_db -t -c "SELECT * FROM pg_extension WHERE extname = 'pg_pathman' AND extversion = '1.3';"
ОШИБКА:  pg_pathman's Pl/PgSQL frontend is incompatible with its shared library
ПОДРОБНОСТИ:  consider performing an update procedure
ПОДСКАЗКА:  pg_pathman will be disabled to allow you to resolve this issue
$ echo $?
1
```

```
$ psql target_db -t -c 'ALTER EXTENSION pg_pathman UPDATE TO "1.3"; SET pg_pathman.enable = t'
ОШИБКА:  pg_pathman's Pl/PgSQL frontend is incompatible with its shared library
ПОДРОБНОСТИ:  consider performing an update procedure
ПОДСКАЗКА:  pg_pathman will be disabled to allow you to resolve this issue
$ echo $?
1
```

**Comment by funbringer** [comment]
Hi @Envek,

> **Actual behavior**
> These commands fails with error with exit code 1 and does nothing:

Thank you for such a thorough explanation :) I made it work that way, for a reason, of course. We tend to change definitions of functions / tables / other SQL objects from time to time, which means that pl/PgSQL part of pg_pathman needs to be updated. That's why you have to apply updates via `ALTER EXTENSION` in the first place.

The `pg_pathman.so` lib can't read deprecated table definitions etc (that would be a huge waste of time), so we had to add a mechanism that would immediately halt the initialization process. The easiest way to do that is to disable pg_pathman and throw an ERROR. This way the users will know what happened, and what they are expected to do in order to "just fix the damn DB". Problem is, literally ANY query (first in session) triggers this ERROR, even the one that begins with `SET`.

This is definitely not a frequent execution path, so we don't really want to bother. Time is running short, and we have to solve a bunch of tasks with a higher priority. You have to understand us.

We'll try to come up with a clean solution to your problem, but honestly, I'm sure that it can be solved with an ugly `do-exception` block like this:

```plpgsql
do language plpgsql
$$
begin
    perform from pg_class limit 1; -- dummy query that triggers error
exception when others then
    EXECUTE 'ALTER EXTENSION pg_pathman UPDATE TO "1.3"; SET pg_pathman.enable = t';
end;
$$;
```

**Comment by Envek** [comment]
Thank you for rapid reply.

Yes, I understand that this is absolutely not crucial and not urgent (even for me), and any workaround will be fine.

But unfortunately `do-exception` blocks doesn't catch this error (is this exception at all?)

**Comment by funbringer** [comment]
> But unfortunately do-exception blocks doesn't catch this error

You're right, I forgot that we emit this ERROR before plpgsql's evaluation function takes place.

I think we'll disable this check for `SET pg_pathman.enable = f` and probably for `ALTER EXTENSION pg_pathman UPDATE`. However, this will be fixed only in 1.4+.

**Comment by funbringer** [comment]
This has been fixed in 1.4.
